core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java

```diff
@@ -0,0 +1,652 @@
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.platform.launcher.Launcher;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.PostDiscoveryFilter;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+import org.junit.platform.launcher.core.LauncherConfig;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+import org.opentest4j.TestAbortedException;
+
+import io.quarkus.bootstrap.app.CuratedApplication;
+import io.quarkus.deployment.dev.ClassScanResult;
+import io.quarkus.deployment.dev.DevModeContext;
+import io.quarkus.dev.console.QuarkusConsole;
+import io.quarkus.dev.testing.TracingHandler;
+
+/**
+ * This class is responsible for running a single run of JUnit tests.
+ */
+public class JunitTestRunner {
+
+    private static final Logger log = Logger.getLogger(JunitTestRunner.class);
+    private final long runId;
+    private final DevModeContext devModeContext;
+    private final CuratedApplication testApplication;
+    private final ClassScanResult classScanResult;
+    private final TestClassUsages testClassUsages;
+    private final TestState testState;
+    private final List<TestRunListener> listeners;
+    List<PostDiscoveryFilter> additionalFilters;
+    private final Set<String> includeTags;
+    private final Set<String> excludeTags;
+    private final Pattern include;
+    private final Pattern exclude;
+    private final boolean displayInConsole;
+
+    private volatile boolean testsRunning = false;
+    private volatile boolean aborted;
+    private volatile boolean paused;
+
+    public JunitTestRunner(Builder builder) {
+        this.runId = builder.runId;
+        this.devModeContext = builder.devModeContext;
+        this.testApplication = builder.testApplication;
+        this.classScanResult = builder.classScanResult;
+        this.testClassUsages = builder.testClassUsages;
+        this.listeners = builder.listeners;
+        this.additionalFilters = builder.additionalFilters;
+        this.testState = builder.testState;
+        this.includeTags = new HashSet<>(builder.includeTags);
+        this.excludeTags = new HashSet<>(builder.excludeTags);
+        this.include = builder.include;
+        this.exclude = builder.exclude;
+        this.displayInConsole = builder.displayInConsole;
+    }
+
+    public void runTests() {
+        long start = System.currentTimeMillis();
+        ClassLoader old = Thread.currentThread().getContextClassLoader();
+        try {
+
+            ClassLoader tcl = testApplication.createDeploymentClassLoader();
```

Comment:
It doesn't look like this deployment CL is closed, is it? May lead to jars not being closed and locking issues on windows.